### PR TITLE
Feature: prewitt operator

### LINKED ISSUE #204: [Feature] - Add Prewitt edge detection operator
**Is your feature request related to a problem?**

ImageLab has Sobel & Scharr edge detection operators but is missing the Prewitt operator. For a teaching platform, having the full set of classical gradient based edge detectors is important so students can compare their behaviour side by side in a pipeline.

**Describe the solution you'd like**

A new Prewitt operator in the Sobel Derivatives category that applies two 3x3 gradient kernels to compute edge magnitude, following the same full-stack pattern as the existing Sobel and Scharr operators.

I will be working on this issue.

### PR DESCRIPTION
**Pull request by xxivani**
## Description
Adds a Prewitt edge detection operator to the Sobel Derivatives category. The backend applies two 3x3 gradient kernels (`Gx = [[-1,0,1],[-1,0,1],[-1,0,1]]`, `Gy = Gx^T`) via `cv2.filter2D` and returns the gradient magnitude as a uint8 image. Colour images are automatically converted to grayscale.

Fixes #204 

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?
Ran the pipeline with a Prewitt block on both grayscale and colour images and confirmed correct edge detection output. Compared output against Sobel to verify expected differences in edge strength and weight distribution.

- [x] Existing tests pass
- [ ] New tests added
- [x] Manual testing

## Checklist
- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [x] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally

### CODE REVIEW

File: imagelab-backend/app/operators/sobel_derivatives/prewitt_operator.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    def compute(self, image: np.ndarray) -> np.ndarray:
+        if image.dtype != np.uint8:
+            raise ValueError(f"PrewittOperator expects a uint8 image, got dtype={image.dtype}.")
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY) if len(image.shape) == 3 else image
+
+        grad_x = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_X)
+        grad_y = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_Y)
+
+        magnitude = np.sqrt(grad_x**2 + grad_y**2)
+        return np.clip(magnitude, 0, 255).astype(np.uint8)
```
Comment (ivantha) [inline imagelab-backend/app/operators/sobel_derivatives/prewitt_operator.py]: **[HIGH]** No unit tests added for new operator

The PR checklist explicitly notes that no new tests were added. For a new image processing operator, unit tests should cover at minimum:
- Valid `uint8` grayscale input → correct output shape and dtype
- Valid `uint8` BGR colour input → correct grayscale-converted edge output
- Non-`uint8` input (e.g. `float32`) → `ValueError` raised with correct message
- Edge case: uniform (blank) image → all-zero magnitude output
- Edge case: single sharp edge → non-zero magnitude at boundary

Without tests, regressions in the kernel values or magnitude calculation are invisible to CI.

**Suggestion:**
```
```python
# tests/operators/sobel_derivatives/test_prewitt_operator.py
import numpy as np
import pytest
from app.operators.sobel_derivatives.prewitt_operator import PrewittOperator

op = PrewittOperator()

def test_output_dtype_and_shape_grayscale():
    img = np.zeros((64, 64), dtype=np.uint8)
    result = op.compute(img)
    assert result.dtype == np.uint8
    assert result.shape == (64, 64)

def test_output_shape_bgr():
    img = np.zeros((64, 64, 3), dtype=np.uint8)
    result = op.compute(img)
    assert result.shape == (64, 64)

def test_raises_on_non_uint8():
    img = np.zeros((64, 64), dtype=np.float32)
    with pytest.raises(ValueError, match="uint8"):
        op.compute(img)

def test_blank_image_gives_zero_magnitude():
    img = np.zeros((64, 64), dtype=np.uint8)
    result = op.compute(img)
    assert result.max() == 0

def test_vertical_edge_detected():
    img = np.zeros((32, 32), dtype=np.uint8)
    img[:, 16:] = 255  # sharp vertical edge
    result = op.compute(img)
    assert result[:, 15].mean() > 50  # strong response at edge column
```
```

File: imagelab-backend/app/operators/sobel_derivatives/prewitt_operator.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+_KERNEL_X = np.array([[-1, 0, 1], [-1, 0, 1], [-1, 0, 1]], dtype=np.float64)
+_KERNEL_Y = np.array([[-1, -1, -1], [0, 0, 0], [1, 1, 1]], dtype=np.float64)
+
+
+class PrewittOperator(BaseOperator):
+    """Applies the Prewitt operator to detect edges using horizontal and vertical gradient kernels."""
+
+    def compute(self, image: np.ndarray) -> np.ndarray:
+        if image.dtype != np.uint8:
+            raise ValueError(f"PrewittOperator expects a uint8 image, got dtype={image.dtype}.")
```
Comment (ivantha) [inline imagelab-backend/app/operators/sobel_derivatives/prewitt_operator.py]: **[MEDIUM]** Channel count not validated before BGR→grey conversion

The guard `len(image.shape) == 3` only checks that the array is 3-dimensional (i.e. has a channel axis). It does **not** verify that the channel count is 3. Two concrete failure modes exist:

1. **RGBA image** `(H, W, 4)` — `len(shape) == 3` is `True`, but `cv2.COLOR_BGR2GRAY` expects exactly 3 channels and raises:
   ```
   cv2.error: (-215:Assertion failed) scn == 3 || scn == 4
   ```
   (OpenCV *does* accept 4-channel BGR→GREY, but only with `cv2.COLOR_BGRA2GRAY`, not `cv2.COLOR_BGR2GRAY`.)

2. **Squeezed single-channel image stored as 3-D** `(H, W, 1)` — same assertion failure.

The fix is to branch on the actual channel count (`image.shape[2]`) rather than the number of dimensions.

**Suggestion:**
```
```python
if image.ndim == 2:
    gray = image
elif image.ndim == 3:
    channels = image.shape[2]
    if channels == 3:
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    elif channels == 4:
        gray = cv2.cvtColor(image, cv2.COLOR_BGRA2GRAY)
    elif channels == 1:
        gray = image[:, :, 0]
    else:
        raise ValueError(
            f"PrewittOperator expects a 2-D or 3-D image with 1/3/4 channels, "
            f"got shape={image.shape}."
        )
else:
    raise ValueError(
        f"PrewittOperator expects a 2-D or 3-D array, got ndim={image.ndim}."
    )
```
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This is a well-structured, focused PR that cleanly integrates a new Prewitt edge detection operator across the full stack (backend implementation, operator registry, frontend block definition, categories list, and documentation). The kernel values and magnitude formula are mathematically correct, and the code follows existing patterns in the codebase. The main concerns are: (1) the absence of unit tests for a new operator, (2) a channel-count validation bug that would cause a runtime crash on RGBA or `(H,W,1)` input images, and (3) an undocumented BGR colour-order assumption. None of the issues affect existing functionality. Addressing the channel validation bug and adding tests before merge is recommended.

**Suggested Version Bump:** `minor`

### ⚠️ Breaking Changes

- **[INFO]** No breaking changes — purely additive feature: **Previous behavior**: The `sobelderivatives_*` registry and frontend category contained only Sobel and Scharr entries.

**New behavior**: A third entry `sobelderivatives_prewittoperator` is added to the registry, frontend categories list, block definitions, and operator documentation.

**Who is affected**: No existing consumers are affected. The new registry key is novel and does not collide with any existing key.

**Migration path**: None required. This is a backwards-compatible additive change.
  > **Suggestion:** No action needed. Suggested semver bump: `minor`.


### PR Description Issues

- **[LOW]** No tests added — not addressed in description body: The PR checklist correctly leaves '- [ ] New tests added' unchecked. However, the description body does not explain *why* tests were omitted, nor does it note which test file or test strategy is planned as a follow-up. For a reviewer or project maintainer this raises an open question: is testing considered out of scope for this PR intentionally, or was it forgotten?

A one-sentence rationale or a follow-up issue reference (e.g. 'Tests will be added in #205') would close this ambiguity.
  > **Suggestion:** Add to the **How Has This Been Tested?** section:
> Unit tests for `PrewittOperator` (dtype validation, grayscale/colour input, blank/edge inputs) will be tracked in follow-up issue #XXX. Omitted here to keep the PR focused on the feature implementation.
- **[INFO]** Screenshots section is empty: The **Screenshots (if applicable)** section is present but contains no content. For a visual image processing feature, before/after screenshots comparing Prewitt, Sobel, and Scharr outputs would be highly valuable for reviewers and help validate the visual correctness of the implementation without running the pipeline locally.
  > **Suggestion:** Add side-by-side output images (e.g. Sobel vs Prewitt on a standard test image such as `lena` or a geometric shapes image) or link to them. Even a single screenshot of the Prewitt output in the UI pipeline would significantly improve reviewer confidence.

**Comment by ivantha** [comment]
Reviewed. Adds a Prewitt edge detection operator with a thorough full-stack implementation: immutable 3×3 kernel arrays, strict uint8 dtype validation, grayscale conversion for 1/3/4-channel inputs, `np.hypot` for gradient magnitude, registry entry, frontend block, category listing, and `operatorDocs` entry with formula. Merging via rebase.

**Comment by ivantha** [comment]
Thanks for the PR! It currently has merge conflicts with main (or is behind). Could you rebase onto the latest main and push the updated branch? Happy to merge once it's auto-rebasable.

**Comment by xxivani** [comment]
Hi @ivantha, I've pushed the updated branch!
